logical: Add replanning logic to LDR

navsetlur opened the pull request:
Currently there's no replanning logic for logical replication, which means
there is no handling for changes to the cluster (e.g. nodes being added or failing).
This change uses the PhysicalPlanChecker, which at regular intervals compares the
initial plan with a more up to date plan and throws an error whne they differ too much,
leading to a replan for the actual logical replication.

Release note: none
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-39122

navsetlur committed:
logical: Add replanning logic to LDR

Currently there's no replanning logic for logical replication, which means
there is no handling for changes to the cluster (e.g. nodes being added or failing).
This change uses the PhysicalPlanChecker, which at regular intervals compares the
initial plan with a more up to date plan and throws an error whne they differ too much,
leading to a replan for the actual logical replication.

Release note: none
Part of: https://cockroachlabs.atlassian.net/browse/CRDB-39122

navsetlur committed:
logical: Add ReplanCount to logical metrics

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126282)
<!-- Reviewable:end -->

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
I think rather than making the heartbeat sender in here, you'll want to make it outside of the func so that then you can have another goroutine that you pass to GoAndWait which will do something like:

```
	return heartbeatSender.Wait()
```

But this can happen in a different commit or PR.

navsetlur replied:
This I'll handle in a different PR

stevendanna reviewed (commented):
Thanks for moving this forward. I took a brief look at the draft and left some comments.

navsetlur commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
I like the idea of capturing all the fields in the struct, would make it easier to read and understand. 

> Have two methods handing off this struct:

Not the biggest fan of this as both of these methods would be identical except for part that updates the frontier and initial plan at the beginning. What I was thinking instead is a pseudo-constructor that sets all the captured variables and a generatePlan function that still sets the initial plan if necessary and returns the context. I can still make a it a top level function so it's easier to understand.

stevendanna replied:
That said, do what makes sense to you.

msbutler commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
what other changes were you planning here that would make LDR's replanning policy different to PCRs here? If none, i suggest refactoring the PCR code to become more extensible, allowing PCR and LDR to use common code.

navsetlur replied:
moved measurePlanChange into distsql_plan_bulk.go and separated out getNodes. Thoughts?

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
This err declaration shouldn't be needed.

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
```diff
@@ -251,51 +427,17 @@ func (r *logicalReplicationResumer) ingest(
 		physicalPlan,
 		execCfg.InternalDB,
-		jobID)
+		p.jobID)
```
I think we'll want to move this StorePlanDiagram out of this function so that we only store the initial plan diagram and not every one that we re-generate.

stevendanna commented on pkg/ccl/crosscluster/settings.go:
```diff
@@ -109,3 +109,20 @@ var ReplicateSpanConfigsEnabled = settings.RegisterBoolSetting(
 )
+
+var LogicalReplanThreshold = settings.RegisterFloatSetting(
+	settings.SystemOnly,
```
ApplicationLevel

navsetlur replied:
Why Application over System? Is this something we want users tinkering around with?

stevendanna replied:
It has more to do with what part of the system the setting is controlling. Typically anything up at the SQL level should be ApplicationLevel whereas anything down in KV is often System level. One way to think about it is whether or not we would want a user in Serverless to be able to modify the setting.

In PCR a lot of the settings are SystemLevel only because PCR is system-tenant-only.  Here, I think most of our settings should probably be ApplicationLevel.

stevendanna commented on pkg/ccl/crosscluster/settings.go:
```diff
@@ -109,3 +109,20 @@ var ReplicateSpanConfigsEnabled = settings.RegisterBoolSetting(
+)
+
+var LogicalReplanFrequency = settings.RegisterDurationSetting(
+	settings.SystemOnly,
```
ApplicationLevel

navsetlur commented on pkg/ccl/crosscluster/logical/logical_replication_job.go:
Feel like it doesn't hurt, makes debug logging easier in the future. But I'm not sure what common practice is around this.

msbutler replied:
In general, ctx is passed as function arg not in a struct. There's a good discussion for why in the tech note FAQ https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/contexts.md#faqs

navsetlur replied:
ah got it, will fix.

msbutler commented on pkg/ccl/crosscluster/logical/metrics.go:
```diff
@@ -82,6 +82,12 @@ var (
 		Measurement: "Events",
 		Unit:        metric.Unit_COUNT,
 	}
+	metaDistSQLReplanCount = metric.Metadata{
```
I think you need to actually increment the new metric? Also, could you format the commit message to CRL standards? Eg:

```
crosscluster/logical: add replan metric

This patch adds a new metric blah
```

navsetlur replied:
I increment the metric in the following commit, is that what you mean? Or did I miss a setup step.

Thanks, I forget to change the header for the commit. As for adding the whole body, if it's a small enough change do I need to explain all that? i was following your example here: https://github.com/cockroachdb/cockroach/pull/126016/commits

navsetlur committed:
crosscluster: Refactor measurePlanChange to be reusable

This code can now be used by both physical and logical replication
to measure distSQL plan divergence.

Epic: none
Release note: none

navsetlur commented:
Added 2 followups:
https://cockroachlabs.atlassian.net/browse/CRDB-39982
https://cockroachlabs.atlassian.net/browse/CRDB-39983

navsetlur commented:
> NB: i would not merge this until the test you added passed extended CI (at least under the stress build. ) https://teamcity.cockroachdb.com/viewLog.html?buildId=15929931&buildTypeId=Cockroach_BazelExtendedCi

Thanks Michael, will make sure to double check

navsetlur commented:
bors r= stevendanna
